internal/httputil/httpapi.go

```diff
@@ -89,6 +115,56 @@ func MakeAuthAPI(
 	return MakeExternalAPI(metricsName, h)
 }
 
+func checkConsent(ctx context.Context, userID string, userAPI userapi.UserInternalAPI, userConsentCfg config.UserConsentOptions) error {
+	localPart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return nil
+	}
+	// check which version of the policy the user accepted
+	res := &userapi.QueryPolicyVersionResponse{}
+	err = userAPI.QueryPolicyVersion(ctx, &userapi.QueryPolicyVersionRequest{
+		LocalPart: localPart,
+	}, res)
+	if err != nil {
+		return nil
+	}
+
+	// user hasn't accepted any policy, block access.
+	if userConsentCfg.Version != res.PolicyVersion {
+		uri, err := getConsentURL(userID, userConsentCfg)
+		if err != nil {
+			return jsonerror.Unknown("unable to get consent URL")
+		}
+		msg := &bytes.Buffer{}
+		c := struct {
+			ConsentURL string
+		}{
+			ConsentURL: uri,
+		}
+		if err = userConsentCfg.TextTemplates.ExecuteTemplate(msg, "blockEventsError", c); err != nil {
+			logrus.Infof("error consent message: %+v", err)
+			return jsonerror.Unknown("unable to get consent URL")
```

Comment:
Fix up the error messages on this please. Failure to execute template would be better.